multicluster/controllers/multicluster/node_controller.go

```diff
@@ -73,65 +78,193 @@ func NewNodeReconciler(
+				ObjectMeta: metav1.ObjectMeta{
+					Namespace: r.namespace,
+					Name:      existingGWName},
+			}
+			err := r.Client.Delete(ctx, staleGateway, &client.DeleteOptions{})
+			if err != nil && !apierrors.IsNotFound(err) {
+				return err
+			}
+		} else {
+			r.activeGateway = existingGWName
+		}
+	}
+	for _, n := range nodeList.Items {
+		if _, isGW := n.Annotations[common.GatewayAnnotation]; isGW {
+			r.gatewayCandidates[n.Name] = true
+		}
+	}
+	return nil
+}
+
+func (r *NodeReconciler) updateActiveGateway(ctx context.Context, newGateway *mcsv1alpha1.Gateway) error {
+	existingGW := &mcsv1alpha1.Gateway{}
+	// TODO: cache might be stale. Need to revisit here and other reconcilers to
+	// check if we can improve this with 'Owns' or other methods.
+	if err := r.Client.Get(ctx, types.NamespacedName{Name: newGateway.Name, Namespace: r.namespace}, existingGW); err != nil {
+		return client.IgnoreNotFound(err)
+	}
+	existingGW.GatewayIP = newGateway.GatewayIP
+	existingGW.InternalIP = newGateway.InternalIP
+	if err := r.Client.Update(ctx, existingGW, &client.UpdateOptions{}); err != nil {
+		return err
+	}
+	return nil
+}
+
+// recreateActiveGateway will delete existing Gateway CR and create a new Gateway
+// from the pool of Gateway candidates.
+func (r *NodeReconciler) recreateActiveGateway(ctx context.Context, gateway *mcsv1alpha1.Gateway) error {
+	err := r.Client.Delete(ctx, gateway, &client.DeleteOptions{})
+	if err == nil || apierrors.IsNotFound(err) {
+		r.activeGateway = ""
+		// Check remaining Gateway's candidates and create a new Gateway.
+		newGateway, err := r.getValidGatewayFromCandidates()
+		if err != nil {
+			return err
+		}
+		if newGateway != nil {
+			return r.createGateway(newGateway)
+		}
+		return nil
+	}
+	return err
+}
+
+// getValidGatewayFromCandidates picks a valid Node from Gateway candidates and
+// creates a Gateway. It returns no error if no good Gateway candidate.
+func (r *NodeReconciler) getValidGatewayFromCandidates() (*mcsv1alpha1.Gateway, error) {
+	var activeGateway *mcsv1alpha1.Gateway
+	var internalIP, gwIP string
+	var err error
+	gatewayNode := &corev1.Node{}
+	if len(r.gatewayCandidates) == 0 {
+		return nil, nil
+	}
+	for name := range r.gatewayCandidates {
+		if err = r.Client.Get(ctx, types.NamespacedName{Name: name}, gatewayNode); err == nil {
+			if !isReadyNode(gatewayNode) {
+				continue
+			}
+			if internalIP, gwIP, err = r.getGatawayNodeIP(gatewayNode); err != nil {
+				klog.V(2).ErrorS(err, "Node has no valid IP", "node", gatewayNode.Name)
+				continue
+			}
+			activeGateway = &mcsv1alpha1.Gateway{
+				ObjectMeta: metav1.ObjectMeta{
+					Name:      gatewayNode.Name,
+					Namespace: r.namespace,
+				},
+				GatewayIP:  gwIP,
+				InternalIP: internalIP,
+			}
+			klog.InfoS("Found available Gateway candidate", "node", gatewayNode.Name)
+			return activeGateway, nil
+		}
+		if apierrors.IsNotFound(err) {
+			delete(r.gatewayCandidates, name)
```

Comment:
I feel better not to do such changes in this get func. Let us leave it to the Node deletion event.

Reply:
Removed